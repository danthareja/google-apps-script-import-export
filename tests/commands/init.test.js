const fs = require("fs");
const path = require("path");
const process = require("process");
const { promisify } = require("util");

const unlinkAsync = promisify(fs.unlink);
const rimraf = promisify(require('rimraf'));

const fakeClientSecretPath = "/tmp/.fakeClientSecret";
const defaults = require("../../lib/defaults");

// TODO export in a test helpers file (folder?)
let spiedConsole = jest.spyOn(console, "log").mockImplementation(() => {});

const workingDirectory = "/tmp/gapps-init-test";

const setup = () => {
  fs.mkdirSync(workingDirectory);
  process.chdir(workingDirectory);
}

const teardown = async () => {
  await rimraf(workingDirectory);

  jest.resetAllMocks();
};

beforeEach(setup);
afterEach(teardown);

test("should check the provided fileId is not a project key", async () => {
  const init = require("../../lib/commands/init");
  await init("MFQVRDQbuEJFkzP0Kafn0fVLQHf-4KJyd", {});
  expect(spiedConsole.mock.calls[0][0]).toMatchSnapshot();
});

test("should fail if a config file already exists", async () => {
  const init = require("../../lib/commands/init");

  const configPath = path.join(workingDirectory, defaults.CONFIG_NAME);
  console.log(configPath);

  // TODO if the JSON is malformed, we get an "unexpected end of json input"
  fs.writeFileSync(configPath, '{}');

  try {
    await init("1NS2GWZQtgo1rxPuA13fhagxtfs95M3dE_dbtQ3GYAEiXGBzYZreSUhUu", {});
  } catch(e) {
    expect(e).toMatchSnapshot();
  }
})

test("should clone the project", async () => {
  const init = require("../../lib/commands/init");

  // Test project: https://script.google.com/macros/d/1NS2GWZQtgo1rxPuA13fhagxtfs95M3dE_dbtQ3GYAEiXGBzYZreSUhUu/edit?usp=drive_web&splash=yes
  const fileId = '1NS2GWZQtgo1rxPuA13fhagxtfs95M3dE_dbtQ3GYAEiXGBzYZreSUhUu';

  await init(fileId, {});

  const writtenConfigFile = fs.readFileSync(path.join(workingDirectory, 'gapps.config.json'));
  const parsedConfig = JSON.parse(writtenConfigFile);

  // Make sure a config file was created with the correct file id:
  expect(parsedConfig.fileId).toEqual(fileId);

  // Grab the file in the src folder to make sure it was correctly downloaded:
  const codeFile = fs.readFileSync(path.join(workingDirectory, "src/Code.js")).toString();
  expect(codeFile).toMatchSnapshot();
});
