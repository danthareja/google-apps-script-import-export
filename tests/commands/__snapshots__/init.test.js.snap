// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should check the provided fileId is not a project key 1`] = `
"[31m
Script file ID error.[39m
It looks like you are passing in a Project Key, from \\"File --> Project properties\\",rather than a Drive File ID.
You will find the Drive File ID in the script's URL:
https://script.google.com/a/google.com/d/[32m__DRIVE_FILE_ID__[39m/edit.
"
`;

exports[`should clone the project 1`] = `
"// Project for testing node-google-apps-script

function myFunction() {
  
}
"
`;

exports[`should fail if a config file already exists 1`] = `"Config already exists. Cowardly refusing to overwrite."`;
